File: Using multiple SQL-Server databases in a solution/[C#]-Using multiple SQL-Server databases in a solution/C#/Demo_cs/Classes/Operations.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Data;
using System.Data.SqlClient;

namespace Demo_cs
{
    public class Operations
    {
        /// <summary>
        /// SQL-Server server name
        /// </summary>
        private string ServerName;
        /// <summary>
        /// Connection string to the customer database
        /// </summary>
        private string ConnectionString;
        /// <summary>
        /// Connection string to the City and State database
        /// </summary>
        private string ReferenceConnectionString;
        public bool HasErrors { get; set; }
        public string ExceptionMessage { get; set; }
        public DataTable CustomersTable;
        public List<State> States;
        public List<City> Cities;

        public Operations()
        {
            /*
             * IMPORTANT: You must change the server name to your server name.
             */
ServerName = "KARENS-PC";

ConnectionString = $"Data Source={ServerName};Initial Catalog=CustomersForCodeSample;Integrated Security=True";
ReferenceConnectionString = $"Data Source={ServerName};Initial Catalog=CityStateReferences;Integrated Security=True";

        }
        /// <summary>
        /// Return all states from the database into a list
        /// </summary>
        /// <returns></returns>
        public bool GetStates()
        {
            var success = false;
            States = new List<State>();

            using (SqlConnection cn = new SqlConnection { ConnectionString = ReferenceConnectionString })
            {
                using (SqlCommand cmd = new SqlCommand { Connection = cn })
                {
                    cmd.CommandText = "SELECT StateIdentifier,Name FROM US_States";
                    try
                    {
                        cn.Open();
                        SqlDataReader reader = cmd.ExecuteReader();
                        States.Add(new State { StateIdentifier = 0, Name = "Select one" });
                        while (reader.Read())
                        {
                            States.Add(new State { StateIdentifier = reader.GetInt32(0), Name = reader.GetString(1) });
                        }

                        success = true;
                    }
                    catch (Exception ex)
                    {
                        HasErrors = true;
                        ExceptionMessage = ex.Message;
                    }
                }
            }

            return success;

        }
        /// <summary>
        /// Return all cities from the database with city and state identifiers
        /// </summary>
        /// <returns></returns>
        public bool GetCities()
        {
            var success = false;
            Cities = new List<City>();

            using (SqlConnection cn = new SqlConnection { ConnectionString = ReferenceConnectionString })
            {
                using (SqlCommand cmd = new SqlCommand { Connection = cn })
                {
                    cmd.CommandText = "SELECT CityIdentifier, StateIdentifier, CityName FROM US_Cities";
                    try
                    {
                        cn.Open();
                        SqlDataReader reader = cmd.ExecuteReader();
                        while (reader.Read())
                        {
                            Cities.Add(new City { CityIdentifier = reader.GetInt32(0), StateIdentifier = reader.GetInt32(1), Name = reader.GetString(2) });
                        }

                        Cities = Cities.OrderBy(item => item.Name).ToList();
                        Cities.Insert(0, new City { CityIdentifier = 0, Name = "Select one", StateIdentifier = 0 });
                        success = true;
                    }
                    catch (Exception ex)
                    {
                        HasErrors = true;
                        ExceptionMessage = ex.Message;
                    }
                }
            }

            return success;

        }
        /// <summary>
        /// Return all cutomers using two databases
        /// </summary>
        /// <returns></returns>
        public bool GetCustomers()
        {
            var success = false;
            CustomersTable = new DataTable();
            using (SqlConnection cn = new SqlConnection { ConnectionString = ConnectionString })
            {
                using (SqlCommand cmd = new SqlCommand { Connection = cn})
                {

                    cmd.CommandText = @"
                        SELECT  C.id,
                                C.FirstName ,
                                C.LastName ,
                                C.CityIdentifier ,
                                C.StateIdentifer ,
                                uc.CityName AS City ,
                                us.Name AS State 
                        FROM    Customers AS C
                                LEFT OUTER JOIN CityStateReferences.dbo.US_States AS us ON C.StateIdentifer = us.StateIdentifier
                                LEFT OUTER JOIN CityStateReferences.dbo.US_Cities AS uc ON uc.CityIdentifier = C.CityIdentifier";

                    try
                    {
                        cn.Open();

                        CustomersTable.Load(cmd.ExecuteReader());

                        // field we want but not to show in the user interface
                        CustomersTable.Columns["id"].ColumnMapping = MappingType.Hidden;
                        CustomersTable.Columns["StateIdentifer"].ColumnMapping = MappingType.Hidden;
                        CustomersTable.Columns["CityIdentifier"].ColumnMapping = MappingType.Hidden;

                        // better to perform sort on client side
                        CustomersTable.DefaultView.Sort = "LastName";
                        success = true;
                    }
                    catch (Exception ex)
                    {
                        HasErrors = true;
                        ExceptionMessage = ex.Message;
                    }

                }
            }

            return success;

        }
        /// <summary>
        /// Add a new cutomer
        /// </summary>
        /// <param name="customer"></param>
        /// <returns></returns>
        public bool AddCustomer(ref Customer customer)
        {
            var success = false;
            using (SqlConnection cn = new SqlConnection { ConnectionString = ConnectionString })
            {
                using (SqlCommand cmd = new SqlCommand { Connection = cn })
                {

                    cmd.CommandText = 
                        "INSERT INTO Customers (FirstName,LastName,StateIdentifer,CityIdentifier)  " + 
                        "VALUES (@FirstName,@LastName,@StateIdentifer,@CityIdentifier)";

                    try
                    {
                        cn.Open();

                        cmd.Parameters.AddWithValue("@FirstName", customer.FirstName);
                        cmd.Parameters.AddWithValue("@LastName", customer.LastName);
                        cmd.Parameters.AddWithValue("@StateIdentifer", customer.StateIdentifer);
                        cmd.Parameters.AddWithValue("@CityIdentifier", customer.CityIdentifier);

                        int result = cmd.ExecuteNonQuery();

                        if (result == 1)
                        {
                            cmd.CommandText = "Select @@Identity";
                            customer.id = Convert.ToInt32(cmd.ExecuteScalar());
                            success = true;
                        }
                    }
                    catch (Exception ex)
                    {
                        HasErrors = true;
                        ExceptionMessage = ex.Message;
                    }
                }
            }

            return success;

        }
    }
}

File: Windows forms Master-detail view with addeditdelete on SQL-Server/[C#][VB]-Windows forms Master-detail view with addeditdelete on SQL-Server/C# and VB/WindowsApplication2/Operations.vb
Imports System.Data.SqlClient

Public Class Operations
    Private Server As String = "KARENS-PC"
    Private Catalog As String = "ForumExamples"
    Private ConnectionString As String = $"Data Source={Server};Initial Catalog={Catalog};Integrated Security=True"
    Property ExceptionMessage As String
    Property HasException As Boolean
    Public Function UpdateRecord(ByVal CategoryID As Integer, ByVal CategoryName As String, ByVal CategoryShortName As String, CategoryLocked As Integer, CategoryUsed As Integer) As Boolean

        HasException = False
        Dim success As Boolean = False

        Using cn As New SqlConnection With {.ConnectionString = ConnectionString}
            Dim sql As String =
                <SQL>
                    UPDATE tblToolCategory 
                    SET                        
                         CategoryName = @CategoryName,
                         CategoryShortName = @CategoryShortName,
                         CategoryLocked = @CategoryLocked,
                         CategoryUsed = @CategoryUsed
                    WHERE CategoryID = @CategoryID
                </SQL>.Value

            Using cmd As New SqlCommand With {.Connection = cn, .CommandText = sql}
                cmd.Parameters.AddWithValue("@CategoryName", CategoryName)
                cmd.Parameters.AddWithValue("@CategoryShortName", CategoryShortName)
                cmd.Parameters.AddWithValue("@CategoryLocked", CategoryLocked)
                cmd.Parameters.AddWithValue("@CategoryUsed", CategoryUsed)
                cmd.Parameters.AddWithValue("@CategoryID", CategoryID)

                Try
                    cn.Open()
                    success = cmd.ExecuteNonQuery = 1
                Catch ex As Exception
                    success = False
                    HasException = True
                    ExceptionMessage = ex.Message
                End Try
            End Using
        End Using

        Return success
    End Function
    Public Function InsertRecord(ByRef CategoryID As Integer, ByVal CategoryName As String, ByVal CategoryShortName As String, ByVal CategoryLocked As Integer, ByVal CategoryUsed As Integer) As Boolean

        HasException = False
        Dim success As Boolean = False

        Using cn As New SqlConnection With {.ConnectionString = ConnectionString}
            Dim sql As String =
                <SQL>
                    INSERT INTO tblToolCategory
                        (CategoryName,CategoryShortName,CategoryLocked,CategoryUsed)
                    VALUES
                        (@CategoryName,@CategoryShortName,@CategoryLocked,@CategoryUsed);
                    SELECT CAST(scope_identity() AS int);
                </SQL>.Value

            Using cmd As New SqlCommand With {.Connection = cn, .CommandText = sql}

                cmd.Parameters.AddWithValue("@CategoryName", CategoryName)
                cmd.Parameters.AddWithValue("@CategoryShortName", CategoryShortName)
                cmd.Parameters.AddWithValue("@CategoryLocked", CategoryLocked)
                cmd.Parameters.AddWithValue("@CategoryUsed", CategoryUsed)

                Try
                    cn.Open()
                    CategoryID = CInt(cmd.ExecuteScalar)
                    success = True
                Catch ex As Exception
                    success = False
                    HasException = True
                    ExceptionMessage = ex.Message
                End Try
            End Using
        End Using

        Return success

    End Function
    Public Function GetRecords() As DataTable

        HasException = False
        Dim dt As New DataTable

        Using cn As New SqlConnection With {.ConnectionString = ConnectionString}
            Dim sql As String =
                <SQL>
                    SELECT 
                        CategoryID, 
                        CategoryName, 
                        CategoryShortName, 
                        CAST(CategoryLocked AS INT) AS CategoryLocked, 
                        CAST(CategoryUsed AS INT)   AS CategoryUsed
                    FROM tblToolCategory
                </SQL>.Value
            Using cmd As New SqlCommand With {.Connection = cn, .CommandText = sql}

                Try
                    cn.Open()
                    dt.Load(cmd.ExecuteReader)

                    dt.Columns("CategoryLocked").ReadOnly = False
                    dt.Columns("CategoryUsed").ReadOnly = False

                Catch ex As Exception
                    ExceptionMessage = ex.Message
                    HasException = True
                End Try
            End Using
        End Using
        Return dt
    End Function

    Public Function UpdateOtherDatabase(ByVal id As Integer) As Boolean
        Using cn As New SqlConnection With {.ConnectionString = ConnectionString}
            Using cmd As New SqlCommand With {.Connection = cn}
                cmd.CommandText =
                <SQL>
                    UPDATE  S
                    SET     S.FirstName = b.FirstName
                    FROM    ForumExamples.dbo.Persons1 S
                            INNER JOIN DeleteOptionDatabase.dbo.Persons1 b ON S.id = b.id
                    WHERE   S.id = @id;
                </SQL>.Value

                cmd.Parameters.AddWithValue("@id", id)
                cn.Open()
                Return cmd.ExecuteNonQuery = 1
            End Using
        End Using
    End Function

    ''' <summary>
    ''' Run view ToolCategory
    ''' </summary>
    ''' <returns></returns>
    Public Function GetRecordsFromView() As DataTable

        HasException = False
        Dim dt As New DataTable

        Using cn As New SqlConnection With {.ConnectionString = ConnectionString}
            Using cmd As New SqlCommand With {.Connection = cn, .CommandText = "SELECT * FROM ToolCategory"}

                Try
                    cn.Open()
                    dt.Load(cmd.ExecuteReader)
                Catch ex As Exception
                    ExceptionMessage = ex.Message
                    HasException = True
                End Try
            End Using
        End Using
        Return dt
    End Function

End Class
